File: benchmarks/values/array_shift.go
package main

import (
	"fmt"
	"time"
)

func main() {
	t0 := time.Now()

	const n = 10000000
	a := make([]int, n)
	for i := 0; i < n; i++ {
		a[i] = i % 10
	}

	b := make([]int, 0, n)
	for len(a) > 0 {
		b = append(b, a[0])
		a = a[1:]
	}

	elapsed := time.Since(t0)
	fmt.Printf("length=%d first=%d last=%d\n", len(b), b[0], b[len(b)-1])
	fmt.Printf("elapsed: %.3fs\n", elapsed.Seconds())
}

File: benchmarks/values/array_shift.rs
use std::collections::VecDeque;
use std::time::Instant;

fn main() {
    let t0 = Instant::now();

    const N: usize = 10_000_000;
    let mut a = VecDeque::with_capacity(N);
    for i in 0..N {
        a.push_back((i % 10) as i32);
    }

    let mut b = Vec::with_capacity(N);
    while let Some(v) = a.pop_front() {
        b.push(v);
    }

    let elapsed = t0.elapsed();
    println!("length={} first={} last={}", b.len(), b[0], b[b.len() - 1]);
    println!("elapsed: {:.3}s", elapsed.as_secs_f64());
}
